File: app/components/folio/console/file_placements/multi_picker_fields/placement_component.rb
# frozen_string_literal: true

class Folio::Console::FilePlacements::MultiPickerFields::PlacementComponent < Folio::Console::ApplicationComponent
  bem_class_name :non_unique_file_id

  def initialize(g:, non_unique_file_id: false, placement_key:, embed_input_options: nil, placement_attributes: nil)
    @g = g
    @non_unique_file_id = non_unique_file_id
    @placement_key = placement_key
    @embed_input_options = embed_input_options
    @placement_attributes = placement_attributes || Folio::Console::FilePlacements::MultiPickerFieldsComponent::DEFAULT_PLACEMENT_ATTRIBUTES
  end

  private
    def data
      stimulus_controller("f-c-file-placements-multi-picker-fields-placement",
                          values: {
                            state:,
                            embed: embed?,
                          },
                          action: {
                            "f-c-files-picker:fileDestroyed" => "onFileDestroyed",
                            "f-c-file-placements-multi-picker-fields-placement:highlight" => "onHighlight",
                          })
    end

    def embed_data_with_defaults
      @embed_data_with_defaults ||= @g.object.folio_embed_data || {}
    end

    def embed?
      return @embed if defined?(@embed)
      @embed = embed_data_with_defaults["active"] == true
    end

    def state
      return "filled" if embed?
      @g.object.file_id.blank? ? "loading" : "filled"
    end

    def embed_data_has_errors?
      return @embed_data_has_errors if defined?(@embed_data_has_errors)
      @embed_data_has_errors = @g.object.errors[:folio_embed_data].present?
    end

    def folio_embed_data_input(g)
      opts = {
        as: :embed,
        compact: true
      }

      if @embed_input_options.is_a?(Hash)
        @embed_input_options.each do |key, value|
          opts[key] ||= value
        end
      end

      g.input :folio_embed_data, opts
    end

    def placement_attribute_input_options(attr)
      case attr
      when :description
        { autosize: true, input_html: { rows: 1, data: stimulus_target("description") } }
      when :alt
        { input_html: { data: stimulus_target("alt") } }
      when :title
        { as: :string }
      when :folio_embed_data
        nil
      else
        {}
      end
    end

    def placement_attribute_wrapper_data(attr)
      case attr
      when :alt
        stimulus_target("altWrap")
      when :folio_embed_data
        stimulus_target("embedFieldsWrap")
      else
        nil
      end
    end
end

File: app/components/folio/console/file_placements/multi_picker_fields/placement_component.js
window.Folio.Stimulus.register('f-c-file-placements-multi-picker-fields-placement', class extends window.Stimulus.Controller {
  static values = {
    state: String,
    embed: { type: Boolean, default: false }
  }

  static targets = ['alt', 'description', 'pickerWrap', 'altWrap', 'embedFieldsWrap']

  connect () {
    if (this.stateValue !== 'filled') {
      const filled = this.fillFromParent()

      if (filled) {
        this.stateValue = 'filled'
      } else {
        console.error('Failed to fill from parent! Removing .f-c-file-placements-multi-picker-fields-placement')
        this.element.dispatchEvent(new CustomEvent('f-nested-fields:removeFields', { bubbles: true }))
      }
    }
  }

  disconnect () {
    if (this.highlightTimeout) {
      window.clearTimeout(this.highlightTimeout)
    }
  }

  fillFromParent () {
    const parent = this.element.closest('.f-nested-fields__fields')

    if (parent) {
      if (parent.dataset.file) {
        try {
          const fileJson = parent.dataset.file
          const file = JSON.parse(fileJson)
          this.altTarget.value = file.attributes.alt
          this.descriptionTarget.value = file.attributes.description

          const picker = this.element.querySelector('.f-c-files-picker')

          if (!picker) {
            console.error('Failed to find .f-c-files-picker element')
            return false
          }

          picker.setAttribute('data-f-c-files-picker-serialized-file-json-value', fileJson)

          return true
        } catch (error) {
          console.error('Failed to parse JSON from parent dataset', error)
        }
      } else if (parent.dataset.embed === 'true') {
        for (const disabled of this.embedFieldsWrapTarget.querySelectorAll('[disabled]')) {
          disabled.disabled = false
        }

        this.embedFieldsWrapTarget.hidden = false

        if (this.hasPickerWrapTarget) this.pickerWrapTarget.remove()
        if (this.hasAltWrapTarget) this.altWrapTarget.remove()

        this.embedValue = true

        return true
      }
    }

    return false
  }

  onNonUniqueClick (e) {
    e.preventDefault()

    const input = this.element.querySelector('.f-c-files-picker__input--file_id')
    const fileId = input.value
    const otherInputs = this.element.closest('.f-nested-fields__fields-wrap').querySelectorAll(`.f-c-files-picker__input--file_id[value="${fileId}"]`)

    for (const otherInput of otherInputs) {
      if (otherInput !== input) {
        const placement = otherInput.closest('.f-c-file-placements-multi-picker-fields-placement')

        if (placement) {
          placement.scrollIntoView({ behavior: 'smooth', block: 'center' })
          placement.dispatchEvent(new CustomEvent('f-c-file-placements-multi-picker-fields-placement:highlight'))
          break
        }
      }
    }
  }

  onHighlight (e) {
    this.element.classList.add('f-c-file-placements-multi-picker-fields-placement--highlighted')
    this.highlightTimeout = window.setTimeout(() => {
      this.element.classList.remove('f-c-file-placements-multi-picker-fields-placement--highlighted')
    }, 500)
  }
})
